### PR TITLE
4.3 pattern comprehension explained

### PR DESCRIPTION
**Pull request by martin-neotech**
Better examples for pattern comprehension:

example 1

```
MATCH (a:Person {name: 'Keanu Reeves'})
RETURN [(a)-->(b:Movie) WHERE b.title CONTAINS 'Matrix' | b.released] AS years
```

example 2

```
MATCH (a:Person {name: 'Keanu Reeves'})
WITH [(a)-->(b:Movie) | b.released] AS years
UNWIND years AS year
WITH year ORDER BY year
RETURN COLLECT(year) AS sorted_years
```

### CODE REVIEW

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 1]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
-                |(keanu)-[:ACTED_IN]->(thematrixreloaded),
-                |(keanu)-[:ACTED_IN]->(thereplacements),
-                |(keanu)-[:ACTED_IN]->(thematrix),
-                |(keanu)-[:ACTED_IN]->(thedevilsadvocate)
-
-      """.stripMargin)
+      """CREATE
+        #  (keanu:Person {name: 'Keanu Reeves'}),
+        #  (johnnymnemonic:Movie {title: 'Johnny Mnemonic', released: 1995}),
+        #  (somethingsgottagive:Movie {title: 'Somethings Gotta Give', released: 2003}),
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala:33]:
```suggestion
        #  (somethingsgottagive:Movie {title: 'Something's Gotta Give', released: 2003}),
```
If the apostrophe doesn't mean trouble, it is the correct title of the movie so please keep.

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 2]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
-        |* <<cypher-list-comprehension,List comprehension>>
-        |* <<cypher-pattern-comprehension,Pattern comprehension>>""")
+    p("""* <<cypher-lists-general,Lists in general>>
+        #* <<cypher-list-comprehension,List comprehension>>
+        #* <<cypher-pattern-comprehension,Pattern comprehension>>""".stripMargin('#'))
     note{
-      p(
-        """Information regarding operators such as list concatenation (`+`), element existence checking (`IN`) and access (`[]`) can be found <<query-operators-list, here>>.
-          |The behavior of the `IN` and `[]` operators with respect to `null` is detailed <<cypher-working-with-null, here>>.""".stripMargin)
+      p("""Information regarding operators such as list concatenation (`+`), element existence checking (`IN`) and access (`[]`) can be found <<query-operators-list,here>>.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala]:
```suggestion
      p("""Information regarding operators, such as list concatenation (`+`), element existence checking (`IN`), and access (`[]`) can be found <<query-operators-list,here>>.
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 3]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
         """RETURN [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] AS list""", ResultAssertions((r) => {
           r.toList should equal(List(Map("list" -> List(0, 1, 2, 3, 4, 5, 6, 7, 8, 9))))
         })) {
         resultTable()
       }
-      p(
-        """In our examples, we'll use the <<functions-range,`range`>> function.
-          |It gives you a list containing all numbers between given start and end numbers.
-          |Range is inclusive in both ends.""".stripMargin)
+      p("""In the examples, you will use the <<functions-range,`range`>> function.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala]:
```suggestion
      p("""In the examples, you use the <<functions-range,`range`>> function.
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 4]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
+          #It gives you a list containing all numbers between given start and end numbers.
+          #Range is inclusive in both ends.""".stripMargin('#'))
       p("""To access individual elements in the list, we use the square brackets again.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala]:
```suggestion
      p("""To access individual elements in the list, you can use the square brackets again.
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 5]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
-          |This will extract from the start index and up to but not including the end index.""")
-      query(
-        """RETURN range(0, 10)[3]""", ResultAssertions((r) => {
+          #This will extract from the start index and up to but not including the end index.""".stripMargin('#'))
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala]:
```suggestion
          #This extracts from the start index and up to, but not including, the end index.""".stripMargin('#'))
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 6]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
       }
     }
     section("Pattern comprehension", "cypher-pattern-comprehension") {
-      p(
-        """Pattern comprehension is a syntactic construct available in Cypher for creating a list based on matchings of a pattern.
-          |A pattern comprehension will match the specified pattern just like a normal `MATCH` clause, with predicates just
-          |like a normal `WHERE` clause, but will yield a custom projection as specified.""".stripMargin)
-      p("The following graph is used for the example below:")
+      p("""Pattern comprehension is a syntactic construct available in Cypher for creating a list based on matchings of a pattern.
+          #A pattern comprehension will match the specified pattern just like a normal `MATCH` clause, with predicates just
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala]:
```suggestion
          #A pattern comprehension matches the specified pattern like a normal `MATCH` clause, with predicates
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 7]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
+          #like a normal `WHERE` clause, but will yield a custom projection as specified.""".stripMargin('#'))
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala]:
```suggestion
          #like a normal `WHERE` clause, but yields a custom projection as specified.""".stripMargin('#'))
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 8]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
+      p("The following graph is used for the following pattern comprehension examples:")
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala]:
```suggestion
      p("The following graph is used for the pattern comprehension examples:")
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 9]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
       graphViz()
-      query(
-        """MATCH (a:Person {name: 'Keanu Reeves'})
-          |RETURN [(a)-->(b) WHERE b:Movie | b.released] AS years """, ResultAssertions((r) => {
-          r.toList.head("years").equals(List(1995, 1997, 1999, 2000, 2003, 2003, 2003))
+      p("""The following example returns a list that contains the year when the movies was released.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala]:
```suggestion
      p("""This example returns a list that contains the year when the movies was released.
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 10]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
+          #The pattern matching in the pattern comprehension look for `Matrix` in the movie title and that the node `a` (`Person` node with the name `Keanu Reeves`) has a relationship with the movie.""".stripMargin('#'))
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala]:
```suggestion
          #The pattern matching in the pattern comprehension looks for `Matrix` in the movie title and that the node `a` (`Person` node with the name `Keanu Reeves`) has a relationship with the movie.""".stripMargin('#'))
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 11]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
+      query("""MATCH (a:Person {name: 'Keanu Reeves'})
+              #RETURN [(a)-->(b:Movie) WHERE b.title CONTAINS 'Matrix' | b.released] AS years""".stripMargin('#'), ResultAssertions((r) => {
+          r.toList.head("years").equals(List(1999, 2003, 2003, 2021))
         })) {
         resultTable()
       }
       p("The whole predicate, including the `WHERE` keyword, is optional and may be omitted.")
+
+
+      p("""The following example returns a sorted list that contains years.
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala]:
```suggestion
      p("""This example returns a sorted list that contains years.
```

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 12]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
+          #The pattern matching in the pattern comprehension look for movie nodes that has a relationship with the node `a` (`Person` node with the name `Keanu Reeves`).""".stripMargin('#'))
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala]:
```suggestion
          #The pattern matching in the pattern comprehension looks for movie nodes that has a relationship with the node `a` (`Person` node with the name `Keanu Reeves`).""".stripMargin('#'))
```

**Review by AlexicaWright — COMMENTED** [review]
I haven't built it, so the comments are purely editorial.

File: cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala [hunk 13]
```diff
@@ -27,149 +27,145 @@ class ListsTest extends DocumentingTest {
...
-          |Range is inclusive in both ends.""".stripMargin)
-      p("""To access individual elements in the list, we use the square brackets again.
-          |This will extract from the start index and up to but not including the end index.""")
-      query(
-        """RETURN range(0, 10)[3]""", ResultAssertions((r) => {
+      p("""In the examples, you use the <<functions-range,`range`>> function.
+          #It gives you a list containing all numbers between given start and end numbers.
+          #Range is inclusive in both ends.""".stripMargin('#'))
+      p("""To access individual elements in the list, you can use the square brackets again.
+          #This extracts from the start index and up, to but not including, the end index.""".stripMargin('#'))
```
Comment (AlexicaWright) [inline cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ListsTest.scala]:
```suggestion
          #This extracts from the start index and up to, but not including, the end index.""".stripMargin('#'))
```

**Review by AlexicaWright — COMMENTED** [review]
Just one little comma in the wrong place. If this is because my suggestion, I apologize for the confusion.

**Review by blouerat — APPROVED** [review]

**Review by AlexicaWright — APPROVED** [review]
🎉  Looks good! Thank you @martin-neotech !

**Pull request merged**